packages/apollo-server-core/src/ApolloServer.ts

```diff
@@ -415,17 +415,24 @@ export class ApolloServerBase {
       // a federated schema!
       this.requestOptions.executor = gateway.executor;
 
-      return gateway.load({ apollo: this.apolloConfig, engine: engineConfig })
-        .then(config => config.schema)
-        .catch(err => {
+      return gateway
+        .load({ apollo: this.apolloConfig, engine: engineConfig })
+        .then((config) => {
+          this.toDispose.add(
+            async () => gateway.stop && (await gateway.stop()),
```

Comment:
Nit: I don't think this needs to be `await`ed here since the `toDispose` array is wrapped in an `await`ed `Promise.all` in the server's `stop` fn. Also added an optional chain in my suggestion.
 
```suggestion
            () => gateway.stop?.(),
```

Reply:
Ooh, I totally forgot about `?.()` and was confused about why I couldn't get `gateway?.stop()` to work.

It does still need to be `gateway?.stop?.()` because I added `stop` as an optional method on the `GraphQLService` interface ... just in case there is somebody out there who for some reason has a non-`@apollo/gateway` implementation of `GraphQLService`. (I think in AS3 we should get rid of that abstraction and just have gateway be ApolloGateway.)

I do actually find `async () => await gateway?.stop?.()` to be clearer than `() => gateway?.stop?.()` in that it makes it really obvious that stop is an async function. Plenty of stop functions exist in the Node world that aren't async, and so to me it adds some clarity. Is that weird of me?

Reply:
I guess you told me the same thing last year and I ignored you that time. https://github.com/apollographql/apollo-server/pull/4450#discussion_r466078381

Reply:
You'd only need the first ? `gateway?` if `gateway` was possibly nullish, but we know here that it isn't. It's only the `stop` fn on the `gateway` object that might not exist so just the one is sufficient `async () => await gateway.stop?.()`.

And that's a fine reason to me to keep it around if you prefer!

Reply:
Actually... I'm continuing to find myself uncomfortable with functions that return `ValueOrPromise<void>`. After all "void" means "I'm not going to look at the return value" which is sorta incompatible with "but I need to see if it's a Promise or not". When it's in our public API then we're kinda stuck with it, but toDispose is entirely private, so I'd like to just make it `Set<() => Promise<void>>` by making the one place where we have a sync function just not be sync.

And once you do that, you actually do need the async/await here, since in the "there is no `stop`" case it's not returning a Promise.

Also completely ignore my comment above about needing a question mark after `gateway`, that was wrong.

Reply:
Aren't we only stuck with it as far as compile-time typings go?  Might we consider changing that typing in our public API within v3?  If so, let's make a ticket and put it on the milestone?

Reply:
Yes, though I would really just be tempted to eliminate GraphQLService and change the parameter to `gateway?: ApolloGateway` too.  (Or even just add "executeOperation" and "loadSchema" hooks to the plugin interface and make gateway into a plugin.) https://github.com/apollographql/apollo-server/issues/4919